### DISCUSSION TITLE
MooseFS 3.x Erasure Code Support

### DISCUSSION THREAD

**Discussion by anon314159** [moosefs/moosefs#558] [General]
Hello,

I just started playing around with MooseFS in one of my development environments and absolutely fell in love with it.  From an experience perspective I have deployed multple Ceph and GlusterFS peta-scale clusters over the past six years.  So my familiarity and comfort level with distributed filesystems is pretty high.  I digress, while my initial experience with MooseFS has been very good, one glaring issue pops up and that is the lack of erasure code support makes it unusable in large scale deployments due to the massive waste of space caused by replication of data assigned via storage goals.  While I'm new to MooseFS and haven't played around with the extensive configuration options for storage classes, not enabling or allowing EC in the free versions seems like "pricing yourself out of the market" against other open source distributed file systems.   More to the point, are there any plans of adding erasure codes to the GPL version? Or better yet, am I simply missing something from the configuration options that would enable this feature?

**Comment by rusticek** [comment]
Iam not the right person to answer you if there will be EC in CE edition, but i know that it is not implemented in current MFS 3.x CE, neither metadada high availability. And missing metadata HA is currently spof and the biggest disadvantage (for me) and reason why im will be deploying and testing CEPH. (i tried to get a quotation twice, but with no answer from their business dpt... iam not blaming)
There are some information from developers(?) that MFS 4 CE will have matadata high availability, but noone know when the release will happen and if that mean zero data loss in case of primary mfsmaster master loss.

MFS is great and rock solid piece of software. In our company we are using it since MFS 2.x
We are using all the functionalities, like snaphosts, storage classes, quotas.

**Reply by rafamiga** [reply]
Would you mind to share how you use snapshots? I've never used it myself, so I'm curious.

**Reply by anon314159** [reply]
Hello,

While the CE version doesn't offer true HA (i.e. zero downtime) for the master,  any server in your cluster running the metalogger service can act as the master in the event of a failure.  It simply a matter of backing up the masters configuration file, installing the master service, updating the configuration files on the chunk servers to reflect the new  master and recycling some services.  This can all be done in minutes with ansible or ssh + keys regardless of how many nodes are in your cluster.  Sure, it's annoying but complex problems require elegant solutions.  In short just standup up additional metaloggers and updating the chunk servers over to them in the event your master fails.  Not to knock on a good product but some design decisions seem geared towards gimping the product in favor of monetization schemes (ex. Lack of HA or EC), which makes MooseFS kind of a pay to play / trialware product.

**Comment by eleaner** [comment]
@anon314159 
> Would you mind to share how you use snapshots? I've never used it myself, so I'm curious.

I use snapshots as you would Mac Time Machine.

Imagine I have some important files and I accidentally deleted them.
Sure - there is offsite backup but recovery will take quite some time.
if I have a snapshot available from before the deletion - The files are still there and easily recoverable.
you can schedule them easily with cron

Another very useful property of snapshots is a quick copy of massive folders
say that you have a backup archive stored in mfs. You want to convert your archive to a newest version of the backup format
you snapshot your backup directory and if an update goes really bad, you have an easy and fast way back

Please consider

utilisation of your storage will raise - the chunks are only deleted when the last snapshot they are in is deleted.
utilisation of your ram will raise a lot (depending on the number of nodes in each the snapshot), you have to rotate them periodically otherwise you will starve your master for ram.

**Reply by anon314159** [reply]
Hello, 

I'm not exactly sure how to enable snapshots within MooseFS.  However, I have exported my MooseFS storage as SMB shares and enabled the VSS plugin for file and folder history with snapshots.  

In order to do this, she must add the following lines to your smb.conf for each shares. 

```
   # Enable VSS
   shadow: snapdir = /path/to/snapdir
   vfs objects = shadow_copy2
   shadow: sort = desc
```

**Comment by chogata** [comment]
Please, check out MooseFS version 4 Community Edition - it has EC codes enabled.
